File: mudpro_desktop_app/lib/modules/company_setup/controller/others_controller.dart
// lib/modules/company_setup/controller/others_controller.dart

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:mudpro_desktop_app/api_endpoint/api_endpoint.dart';
import 'package:mudpro_desktop_app/modules/company_setup/model/others_model.dart';

class OthersController {
  final String baseUrl = ApiEndpoint.baseUrl;

  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    'Accept': 'application/json',
  };

  // ============ ACTIVITY APIs ============
  
  Future<Map<String, dynamic>> addActivities(List<ActivityItem> activities) async {
    try {
      if (activities.length == 1) {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addActivity}'),
          headers: _headers,
          body: jsonEncode(activities.first.toJson()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': 'Activity added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add activity',
          };
        }
      } else {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addBulkActivities}'),
          headers: _headers,
          body: jsonEncode(activities.map((e) => e.toJson()).toList()),
        );

        final responseData = jsonDecode(response.body);

        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': '${activities.length} activities added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add activities',
          };
        }
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<List<ActivityItem>> getActivities() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl${ApiEndpoint.getActivities}'),
        headers: _headers,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        final List items = data['data'] ?? [];
        return items.map((e) => ActivityItem.fromJson(e)).toList();
      } else {
        throw Exception('Failed to load activities');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  Future<Map<String, dynamic>> updateActivity(String id, ActivityItem activity) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl${ApiEndpoint.updateActivity}/$id'),
        headers: _headers,
        body: jsonEncode(activity.toJson()),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Activity updated successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to update activity',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<Map<String, dynamic>> deleteActivity(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl${ApiEndpoint.deleteActivity}/$id'),
        headers: _headers,
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Activity deleted successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to delete activity',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  // ============ ADDITION APIs ============
  
  Future<Map<String, dynamic>> addAdditions(List<AdditionItem> additions) async {
    try {
      if (additions.length == 1) {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addAddition}'),
          headers: _headers,
          body: jsonEncode(additions.first.toJson()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': 'Addition added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add addition',
          };
        }
      } else {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addBulkAdditions}'),
          headers: _headers,
          body: jsonEncode(additions.map((e) => e.toJson()).toList()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': '${additions.length} additions added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add additions',
          };
        }
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<List<AdditionItem>> getAdditions() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl${ApiEndpoint.getAdditions}'),
        headers: _headers,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        final List items = data['data'] ?? [];
        return items.map((e) => AdditionItem.fromJson(e)).toList();
      } else {
        throw Exception('Failed to load additions');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  Future<Map<String, dynamic>> updateAddition(String id, AdditionItem addition) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl${ApiEndpoint.updateAddition}/$id'),
        headers: _headers,
        body: jsonEncode(addition.toJson()),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Addition updated successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to update addition',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<Map<String, dynamic>> deleteAddition(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl${ApiEndpoint.deleteAddition}/$id'),
        headers: _headers,
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Addition deleted successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to delete addition',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  // ============ LOSS APIs ============
  
  Future<Map<String, dynamic>> addLosses(List<LossItem> losses) async {
    try {
      if (losses.length == 1) {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addLoss}'),
          headers: _headers,
          body: jsonEncode(losses.first.toJson()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': 'Loss added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add loss',
          };
        }
      } else {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addBulkLosses}'),
          headers: _headers,
          body: jsonEncode(losses.map((e) => e.toJson()).toList()),
        );

        final responseData = jsonDecode(response.body);

        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': '${losses.length} losses added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add losses',
          };
        }
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<List<LossItem>> getLosses() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl${ApiEndpoint.getLosses}'),
        headers: _headers,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        final List items = data['data'] ?? [];
        return items.map((e) => LossItem.fromJson(e)).toList();
      } else {
        throw Exception('Failed to load losses');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  Future<Map<String, dynamic>> updateLoss(String id, LossItem loss) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl${ApiEndpoint.updateLoss}/$id'),
        headers: _headers,
        body: jsonEncode(loss.toJson()),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Loss updated successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to update loss',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<Map<String, dynamic>> deleteLoss(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl${ApiEndpoint.deleteLoss}/$id'),
        headers: _headers,
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Loss deleted successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to delete loss',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  // ============ WATER-BASED APIs ============
  
  Future<Map<String, dynamic>> addWaterBased(List<WaterBasedItem> waterBased) async {
    try {
      if (waterBased.length == 1) {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addWaterBased}'),
          headers: _headers,
          body: jsonEncode(waterBased.first.toJson()),
        );

        final responseData = jsonDecode(response.body);

        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': 'Water-based item added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add water-based',
          };
        }
      } else {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addBulkWaterBased}'),
          headers: _headers,
          body: jsonEncode(waterBased.map((e) => e.toJson()).toList()),
        );

        final responseData = jsonDecode(response.body);

        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': '${waterBased.length} water-based items added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add water-based',
          };
        }
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<List<WaterBasedItem>> getWaterBased() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl${ApiEndpoint.getWaterBased}'),
        headers: _headers,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        final List items = data['data'] ?? [];
        return items.map((e) => WaterBasedItem.fromJson(e)).toList();
      } else {
        throw Exception('Failed to load water-based');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  Future<Map<String, dynamic>> updateWaterBased(String id, WaterBasedItem waterBased) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl${ApiEndpoint.updateWaterBased}/$id'),
        headers: _headers,
        body: jsonEncode(waterBased.toJson()),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Water-based updated successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to update water-based',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<Map<String, dynamic>> deleteWaterBased(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl${ApiEndpoint.deleteWaterBased}/$id'),
        headers: _headers,
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Water-based deleted successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to delete water-based',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  // ============ OIL-BASED APIs ============
  
  Future<Map<String, dynamic>> addOilBased(List<OilBasedItem> oilBased) async {
    try {
      if (oilBased.length == 1) {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addOilBased}'),
          headers: _headers,
          body: jsonEncode(oilBased.first.toJson()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': 'Oil-based item added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add oil-based',
          };
        }
      } else {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addBulkOilBased}'),
          headers: _headers,
          body: jsonEncode(oilBased.map((e) => e.toJson()).toList()),
        );

        final responseData = jsonDecode(response.body);

        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': '${oilBased.length} oil-based items added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add oil-based',
          };
        }
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<List<OilBasedItem>> getOilBased() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl${ApiEndpoint.getOilBased}'),
        headers: _headers,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        final List items = data['data'] ?? [];
        return items.map((e) => OilBasedItem.fromJson(e)).toList();
      } else {
        throw Exception('Failed to load oil-based');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  Future<Map<String, dynamic>> updateOilBased(String id, OilBasedItem oilBased) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl${ApiEndpoint.updateOilBased}/$id'),
        headers: _headers,
        body: jsonEncode(oilBased.toJson()),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Oil-based updated successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to update oil-based',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<Map<String, dynamic>> deleteOilBased(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl${ApiEndpoint.deleteOilBased}/$id'),
        headers: _headers,
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Oil-based deleted successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to delete oil-based',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  // ============ SYNTHETIC APIs ============
  
  Future<Map<String, dynamic>> addSynthetic(List<SyntheticItem> synthetic) async {
    try {
      if (synthetic.length == 1) {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addSynthetic}'),
          headers: _headers,
          body: jsonEncode(synthetic.first.toJson()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': 'Synthetic item added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add synthetic',
          };
        }
      } else {
        final response = await http.post(
          Uri.parse('$baseUrl${ApiEndpoint.addBulkSynthetic}'),
          headers: _headers,
          body: jsonEncode(synthetic.map((e) => e.toJson()).toList()),
        );

        final responseData = jsonDecode(response.body);
        
        if (response.statusCode == 201 || response.statusCode == 200) {
          return {
            'success': true,
            'message': '${synthetic.length} synthetic items added successfully',
          };
        } else {
          return {
            'success': false,
            'message': responseData['message'] ?? 'Failed to add synthetic',
          };
        }
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<List<SyntheticItem>> getSynthetic() async {
    try {
      final response = await http.get(
        Uri.parse('$baseUrl${ApiEndpoint.getSynthetic}'),
        headers: _headers,
      );

      if (response.statusCode == 200 || response.statusCode == 201) {
        final data = jsonDecode(response.body);
        final List items = data['data'] ?? [];
        return items.map((e) => SyntheticItem.fromJson(e)).toList();
      } else {
        throw Exception('Failed to load synthetic');
      }
    } catch (e) {
      throw Exception('Error: $e');
    }
  }

  Future<Map<String, dynamic>> updateSynthetic(String id, SyntheticItem synthetic) async {
    try {
      final response = await http.put(
        Uri.parse('$baseUrl${ApiEndpoint.updateSynthetic}/$id'),
        headers: _headers,
        body: jsonEncode(synthetic.toJson()),
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Synthetic updated successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to update synthetic',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }

  Future<Map<String, dynamic>> deleteSynthetic(String id) async {
    try {
      final response = await http.delete(
        Uri.parse('$baseUrl${ApiEndpoint.deleteSynthetic}/$id'),
        headers: _headers,
      );

      final responseData = jsonDecode(response.body);

      if (response.statusCode == 200) {
        return {
          'success': true,
          'message': 'Synthetic deleted successfully',
        };
      } else {
        return {
          'success': false,
          'message': responseData['message'] ?? 'Failed to delete synthetic',
        };
      }
    } catch (e) {
      return {
        'success': false,
        'message': 'Error: $e',
      };
    }
  }
}

File: Mud-pro-Backend/src/controllers/others/others.controller.js
// controllers/others/activity.controller.js
import { Activity , Addition, Loss, WaterBased, OilBased, Synthetic} from "../../modules/others/others.model.js";



export const createActivity = async (req, res) => {
  try {
    const data = await Activity.create(req.body);
    res.status(201).json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const createBulkActivities = async (req, res) => {
  try {
    const activities = req.body;
    if (!Array.isArray(activities) || activities.length === 0) {
      return res.status(400).json({ 
        success: false, 
        message: "Please provide an array of activities" 
      });
    }
    const data = await Activity.insertMany(activities);
    res.status(201).json({ 
      success: true, 
      data, 
      saved: data.length,
      message: `${data.length} activities added successfully`
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const getActivities = async (req, res) => {
  try {
    const data = await Activity.find().sort({ createdAt: -1 });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const updateActivity = async (req, res) => {
  try {
    const data = await Activity.findByIdAndUpdate(req.params.id, req.body, { returnDocument: "after" });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const deleteActivity = async (req, res) => {
  try {
    await Activity.findByIdAndDelete(req.params.id);
    res.json({ success: true, message: "Activity deleted" });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

// controllers/others/addition.controller.js

export const createAddition = async (req, res) => {
  try {
    const data = await Addition.create(req.body);
    res.status(201).json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const createBulkAdditions = async (req, res) => {
  try {
    const additions = req.body;
    if (!Array.isArray(additions) || additions.length === 0) {
      return res.status(400).json({ 
        success: false, 
        message: "Please provide an array of additions" 
      });
    }
    const data = await Addition.insertMany(additions);
    res.status(201).json({ 
      success: true, 
      data, 
      saved: data.length,
      message: `${data.length} additions added successfully`
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const getAdditions = async (req, res) => {
  try {
    const data = await Addition.find().sort({ createdAt: -1 });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const updateAddition = async (req, res) => {
  try {
    const data = await Addition.findByIdAndUpdate(req.params.id, req.body, { returnDocument: "after" });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const deleteAddition = async (req, res) => {
  try {
    await Addition.findByIdAndDelete(req.params.id);
    res.json({ success: true, message: "Addition deleted" });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

// controllers/others/loss.controller.js

export const createLoss = async (req, res) => {
  try {
    const data = await Loss.create(req.body);
    res.status(201).json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const createBulkLosses = async (req, res) => {
  try {
    const losses = req.body;
    if (!Array.isArray(losses) || losses.length === 0) {
      return res.status(400).json({ 
        success: false, 
        message: "Please provide an array of losses" 
      });
    }
    const data = await Loss.insertMany(losses);
    res.status(201).json({ 
      success: true, 
      data, 
      saved: data.length,
      message: `${data.length} losses added successfully`
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const getLosses = async (req, res) => {
  try {
    const data = await Loss.find().sort({ createdAt: -1 });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const updateLoss = async (req, res) => {
  try {
    const data = await Loss.findByIdAndUpdate(req.params.id, req.body, { returnDocument: "after" });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const deleteLoss = async (req, res) => {
  try {
    await Loss.findByIdAndDelete(req.params.id);
    res.json({ success: true, message: "Loss deleted" });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

// controllers/others/waterbased.controller.js

export const createWaterBased = async (req, res) => {
  try {
    const data = await WaterBased.create(req.body);
    res.status(201).json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const createBulkWaterBased = async (req, res) => {
  try {
    const waterBased = req.body;
    if (!Array.isArray(waterBased) || waterBased.length === 0) {
      return res.status(400).json({ 
        success: false, 
        message: "Please provide an array of water-based items" 
      });
    }
    const data = await WaterBased.insertMany(waterBased);
    res.status(201).json({ 
      success: true, 
      data, 
      saved: data.length,
      message: `${data.length} water-based items added successfully`
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const getWaterBased = async (req, res) => {
  try {
    const data = await WaterBased.find().sort({ createdAt: -1 });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const updateWaterBased = async (req, res) => {
  try {
    const data = await WaterBased.findByIdAndUpdate(req.params.id, req.body, { returnDocument: "after" });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const deleteWaterBased = async (req, res) => {
  try {
    await WaterBased.findByIdAndDelete(req.params.id);
    res.json({ success: true, message: "Water-based deleted" });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

// controllers/others/oilbased.controller.js

export const createOilBased = async (req, res) => {
  try {
    const data = await OilBased.create(req.body);
    res.status(201).json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const createBulkOilBased = async (req, res) => {
  try {
    const oilBased = req.body;
    if (!Array.isArray(oilBased) || oilBased.length === 0) {
      return res.status(400).json({ 
        success: false, 
        message: "Please provide an array of oil-based items" 
      });
    }
    const data = await OilBased.insertMany(oilBased);
    res.status(201).json({ 
      success: true, 
      data, 
      saved: data.length,
      message: `${data.length} oil-based items added successfully`
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const getOilBased = async (req, res) => {
  try {
    const data = await OilBased.find().sort({ createdAt: -1 });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const updateOilBased = async (req, res) => {
  try {
    const data = await OilBased.findByIdAndUpdate(req.params.id, req.body, { returnDocument: "after" });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const deleteOilBased = async (req, res) => {
  try {
    await OilBased.findByIdAndDelete(req.params.id);
    res.json({ success: true, message: "Oil-based deleted" });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

// controllers/others/synthetic.controller.js

export const createSynthetic = async (req, res) => {
  try {
    const data = await Synthetic.create(req.body);
    res.status(201).json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const createBulkSynthetic = async (req, res) => {
  try {
    const synthetic = req.body;
    if (!Array.isArray(synthetic) || synthetic.length === 0) {
      return res.status(400).json({ 
        success: false, 
        message: "Please provide an array of synthetic items" 
      });
    }
    const data = await Synthetic.insertMany(synthetic);
    res.status(201).json({ 
      success: true, 
      data, 
      saved: data.length,
      message: `${data.length} synthetic items added successfully`
    });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const getSynthetic = async (req, res) => {
  try {
    const data = await Synthetic.find().sort({ createdAt: -1 });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const updateSynthetic = async (req, res) => {
  try {
    const data = await Synthetic.findByIdAndUpdate(req.params.id, req.body, { returnDocument: "after" });
    res.json({ success: true, data });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};

export const deleteSynthetic = async (req, res) => {
  try {
    await Synthetic.findByIdAndDelete(req.params.id);
    res.json({ success: true, message: "Synthetic deleted" });
  } catch (err) {
    res.status(500).json({ success: false, message: err.message });
  }
};